lib/src/modules/task.cc

```diff
@@ -128,11 +152,146 @@ static TaskCommand getTaskCommand(const fs::path &task_executable)
+// will not fail if the directory already exists.
+static fs::path createCacheDir(const fs::path& task_cache_dir, const std::string& sha256) {
+    auto cache_dir = task_cache_dir / sha256;
+    fs::create_directory(cache_dir);
+    fs::last_write_time(cache_dir, time(nullptr));
+    fs::permissions(cache_dir, TASK_FILE_PERMS | fs::owner_exe | fs::group_exe);
+    return cache_dir;
+}
+
+// Computes the sha256 of the file denoted by path. Assumes that
+// the file designated by "path" exists.
+static std::string calculateSha256(const std::string& path) {
+    const std::streamsize CHUNK_SIZE = 2 << 14;
+    char buffer[CHUNK_SIZE];
+    unsigned char md_value[EVP_MAX_MD_SIZE];
+    unsigned int md_len;
+
+    auto md = EVP_sha256();
+    auto mdctx = EVP_MD_CTX_create();
+    EVP_DigestInit_ex(mdctx, md, nullptr);
+    boost::nowide::ifstream ifs(path);
+    while (ifs.read(buffer, CHUNK_SIZE)) {
+      EVP_DigestUpdate(mdctx, buffer, CHUNK_SIZE);
+    }
+    if (!ifs.eof()) {
+      throw Module::ProcessingError(lth_loc::format("Error while reading {1}", path));
+    }
+    EVP_DigestUpdate(mdctx, buffer, ifs.gcount());
+    EVP_DigestFinal_ex(mdctx, md_value, &md_len);
+    EVP_MD_CTX_destroy(mdctx);
+
+    char md_value_ascii[2*EVP_MAX_MD_SIZE+1] = {'\0'};
+    for (unsigned int i = 0; i < md_len; ++i) {
+      snprintf(md_value_ascii+2*i, 3, "%02x", md_value[i]);
+    }
+    return std::string(md_value_ascii);
+}
+
+static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
+    std::string url = uri.get<std::string>("path");
+    auto params = uri.getWithDefault<lth_jc::JsonContainer>("params", lth_jc::JsonContainer());
+    if (params.empty()) {
+        return url;
+    }
+    auto curl_handle = lth_curl::curl_handle();
+    url += "?";
+    for (auto& key : params.keys()) {
+        auto escaped_key = std::string(
+           lth_curl::curl_escaped_string(curl_handle, key));
+        auto escaped_val = std::string(
+           lth_curl::curl_escaped_string(curl_handle, params.get<std::string>(key)));
+        url += escaped_key + "=" + escaped_val + "&";
+    }
+    return url;
+}
+
+// This method downloads the file at the specified url into the provided path
+// If something goes wrong during the download, a PXP-error should be returned.
+static void downloadTaskFile(const std::vector<std::string>& master_uris,
+                             lth_curl::client& client,
+                             const fs::path& file_path,
+                             const lth_jc::JsonContainer& uri) {    
+    auto url = master_uris[0] + createUrlEndpoint(uri);
+    lth_curl::request req(url);
+    // timeout from connection after one minute, can configure 
+    req.connection_timeout(60000);
+    try {
+        client.download_file(req, file_path.string());
+    } catch (lth_curl::http_file_download_exception& e) {
+        // TODO: There's a chance that curl's temp file (not the temp. task file) might not be removed here.
```

Comment:
The way I was thinking with that one is that we have one "temp" file that we save the downloaded task to ("tempA") which is what's passed in to the downloadTask method. When Leatherman.curl downloads the task, it will create another temp "tempB" to write its contents in chunks and then save it to "tempA" after the download is complete. That's where the two temp files come from. I think it isn't a bad idea to have two, because once we have "tempA" we also want to check that the passed-in sha matches "tempA"'s sha for version control.